sql,tracing: reduce overhead of tracing

tbg opened the pull request:
The below is almost the result - it also included https://github.com/cockroachdb/cockroach/pull/135567, which likely needs a bit more discussion, as well as https://github.com/cockroachdb/cockroach/pull/135682 which merged separately.

```
$ env N=15 T=1000x PKG=./pkg/sql/tests BENCHES=BenchmarkSysbench/SQL/3node/oltp_read_write scripts/bench 3d083fe394a70a45e6b6f95dd8b64ed782ede609 tracing-opt

                                                       old   new
Sysbench/SQL/3node/oltp_read_write-8           5.922m ± 2%   5.629m ± 4%   -4.94% (p=0.004 n=15)
Sysbench/SQL/3node/oltp_read_write-8          2.544Mi ± 0%   2.399Mi ± 1%  -5.70% (p=0.000 n=15)
Sysbench/SQL/3node/oltp_read_write-8           18.98k ± 0%   17.46k ± 1%   -7.96% (p=0.000 n=15)
```

<details>Alloc profiles<summary>Details</summary>
<p>

The alloc profiles below were generated with:

```
rm -f *.pb.gz; ./dev bench --ignore-cache --stream-output --bench-mem ./pkg/sql/tests --filter BenchmarkSysbench/SQL/3node/oltp_read_write --test-args '-test.memprofile=after.pb.gz -test.benchtime=10s -test.cpu=8' -- --test_env 'COCKROACH_TESTING_MEM_PROFILES=/Users/tbg/crdb'

pprof -http :6060 -lines -base BenchmarkSysbench_SQL_3node_oltp_read_write.allocs.pre.pb.gz after.pb.gz
```

The above prevents setup-related allocations from showing up in the allocation profile.

Note that with `sql.txn_stats.sample_rate=0`, the `tracing.` related allocs disappear entirely. We could
consider reducing the sample rate by another factor of 10 to `.1%` (from 1%) or to make it more adaptive.
There is likely little benefit to sampling very frequent fingerprints at 1%.

![after](https://github.com/user-attachments/assets/869ed6c8-5771-49b8-b653-5c05b6cb4892)
![before](https://github.com/user-attachments/assets/038dab4e-956a-4600-bae2-f44e34bd4e3e)

</p>
</details> 

Epic: CRDB-43584

tbg committed:
server: don't force trace span in setupSpanForIncomingRPC

We were creating a trace span for every incoming RPC that
did not have tracing on. This is unnecessary and costly: this call alone
caused 1.3% of allocations in `BenchmarkSysbench/SQL/oltp_write_only`

Epic: none

tbg committed:
kvcoord: avoid EnsureChildSpan in DistSender.

0.3% of allocations in oltp_read_write microbenchmark.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/135146)
<!-- Reviewable:end -->

tbg committed:
sql: remove more EnsureChildSpan

These all showed up in the heap profile.

tbg committed:
stop: do not call `tracing.EnsureX`

tbg committed:
kvserver: avoid unused noop spans

tbg committed:
execinfra: check for noop span before starting span with listeners

If StartSpanCtx is called with a noop span but listeners, it will
upgrade to at least a recording type of RecordingStructured, which
implies creating real trace spans. This is very expensive in terms of
allocations and CPU.

This was around 0.3% of alloc_count directly, but removing it removed
all ~6% of tracing-related allocations (note that the span registry was
disabled in a previous commit), as the majority would be incurred in
creating and collecting child spans.

With the execution stats sample rate set to zero, this removes tracing
from the allocation profiles entirely.

dhartunian commented on pkg/kv/kvclient/kvcoord/txn_coord_sender.go:
```diff
@@ -495,7 +496,7 @@ func (tc *TxnCoordSender) Send(
 	}
 
-	ctx, sp := tc.AnnotateCtxWithSpan(ctx, OpTxnCoordSender)
+	ctx, sp := tracing.ChildSpan(ctx, OpTxnCoordSender)
```
Do we care about the extra context stuff that's pulled in from ambient context?

tbg replied:
That's a good point, there is more stuff in there than I thought. Maybe I should make a change that calls `ChildSpan` instead of `EnsureChildSpan` in `AnnotateCtx`?

tbg replied:
Trying that out, this commit is now replaced with `log: don't call EnsureChildSpan in AnnotateCtx`.

dhartunian commented on pkg/util/stop/stopper.go:
```diff
@@ -474,9 +474,9 @@ func (s *Stopper) RunAsyncTaskEx(ctx context.Context, opt TaskOpts, f func(conte
+		ctx, sp = tracing.ForkSpan(ctx, opt.TaskName)
 	case ChildSpan:
-		ctx, sp = tracing.EnsureChildSpan(ctx, s.tracer, opt.TaskName)
+		ctx, sp = tracing.ChildSpan(ctx, opt.TaskName)
```
These changes are slightly more interesting to me. It seems like a good idea to force async tasks to have spans regardless of the call-site. It seems like the overhead might be worth it in these cases, but perhaps my intuition is wrong.

tbg replied:
It might be slightly misleading that they're called "async tasks", as they are in fact used for spawning short-lived request goroutines in the hot path. For example `DistSender` does this, and I think `kvstreamer` too. Besides, if async tasks have spans, what benefit do we derive from this? I think back in the day we were hoping that we'd use the async span registry to look at what the goroutines are doing, etc, but nowadays I am more inclined to believe that execution traces or eBPF or some combination thereof will fill our observability needs better.

tbg replied:
I think this was introduced here https://github.com/cockroachdb/cockroach/commit/2ad2bee257e78970ce2c457ddd6996099ed6727a - note that there isn't strong motivation other than the intuition that you had - but since we use these for short-running frequent tasks a lot of the time, it makes no sense to have trace spans here. And even if it made sense, we wouldn't be making them visible anywhere, so from that point of view it could only make sense to keep doing this if we were also gearing up to cash in on this information. It doesn't look like anybody is, so I'm very comfortable removing this.

dhartunian commented on pkg/kv/kvserver/replica_application_decoder.go:
```diff
@@ -145,7 +145,7 @@ func (d *replicaDecoder) createTracingSpans(ctx context.Context) {
 			var propSp *tracing.Span
 			// If the client has a trace, put a child into propCtx.
-			if sp := tracing.SpanFromContext(cmd.proposal.Context()); sp != nil {
+			if sp := tracing.SpanFromContext(cmd.proposal.Context()); sp != nil && !sp.IsNoop() {
```
curious how you discovered this case. also lots of allocations for the line below?

tbg replied:
Yep, I basically went through the heap profile looking for all samples going into `tracer.StartSpanCtx`.

tbg replied:
Also I'll admit that this is a bit of a Rube Goldberg machine to me at this point. I'm confident that at some point I did see this line make a difference. But what were the other three things I had already turned off by the time? I'm no longer sure. There's a chance that with all of the other commits, this line doesn't make a difference. But it did at some point, similar to another check I added elsewhere (somewhere in distsql server I think) that at the time made all the difference. Ideally we accept some "spurious" fixes for now because it's quite tedious to re-validate each of them.

dhartunian commented on pkg/util/tracing/tracer.go:
```diff
@@ -179,7 +179,7 @@ var EnableActiveSpansRegistry = settings.RegisterBoolSetting(
 	"trace.span_registry.enabled",
 	"if set, ongoing traces can be seen at https://<ui>/#/debug/tracez",
-	envutil.EnvOrDefaultBool("COCKROACH_REAL_SPANS", true),
+	envutil.EnvOrDefaultBool("COCKROACH_REAL_SPANS", false),
```
Agree this is the right call. I haven't observed much usage of `tracez` in the last year.

dhartunian commented on pkg/sql/execinfra/processorsbase.go:
```diff
@@ -840,7 +840,7 @@ func ProcessorSpan(
 ) (context.Context, *tracing.Span) {
 	sp := tracing.SpanFromContext(ctx)
-	if sp == nil {
+	if sp == nil || sp.IsNoop() {
```
This is quite a subtle catch. Do you think we should unify `nil` and `IsNoop()`? It seems like a footgun that you have to remember to check them together.

tbg replied:
Yeah, this ties in well with my other comment above[^1] - I'm unclear on what the intention of all of this code was and assume what we _want_ has drifted now, too.
We sort-of-mostly moved into a world where there are trace spans everywhere. I think the idea was that you have a noopSpan unless tracing is off? Today I think that's a bad idea too though, since you end up allocating a bunch just to wrap `context.Context`. I don't believe that any version of "always-on tracing" is going to be the future and even if it were, we've been paying all of the cost and getting little benefit. So we'd likely want to figure out what it is we want (for example: no spans by default) and clean up with that goal. That isn't something that I can clean up as a drive-by though, I can only offer the lame suggestion to file a follow-up issue against Obs.

[^1]: https://github.com/cockroachdb/cockroach/pull/135146#discussion_r1847829820

dhartunian reviewed (commented):
Thanks for working through this and for the separate commits!

tbg committed:
log: don't call EnsureChildSpan in AnnotateCtx

As of this commit, we no longer force new spans to
be created in that method if there is no incoming
span to begin with.

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r10, 13 of 13 files at r21, 1 of 1 files at r22, 1 of 1 files at r23, 3 of 3 files at r24, 3 of 3 files at r25, 3 of 3 files at r26, 6 of 6 files at r27, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135146#-:-OC3bmO37e88i6_bUCqL:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @kyle-a-wong, and @michae2)

___
*[`pkg/util/tracing/tracer.go` line 1595 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/135146#-OC3aqpS8Xm7E8-_w3wV:-OC3aqpS8Xm7E8-_w3wW:b-roqalv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9001f92f609d601d8ebc18a97a1c401f62e7cd62/pkg/util/tracing/tracer.go#L1595)):*
> ```Go
> }
> 
> // ChildSpan creates a child span of the current one, if any. Recordings from
> ```

nit: mention the WithParent option that is added, like `EnsureChildSpan` does


<!-- Sent from Reviewable.io -->

tbg committed:
tracing: improve comment on ChildSpan

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135146)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @kyle-a-wong, @mgartner, and @michae2)

___
*[`pkg/util/tracing/tracer.go` line 1595 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/135146#-OC3aqpS8Xm7E8-_w3wV:-OC3tXY6A0b8e6cxeyNu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9001f92f609d601d8ebc18a97a1c401f62e7cd62/pkg/util/tracing/tracer.go#L1595)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: mention the WithParent option that is added, like `EnsureChildSpan` does
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
These all LGTM

dhartunian reviewed (approved):
LGTM. Thanks for answering all my questions. Agree that moving forward we can re-visit assumptions that existing code has made about how tracing should be used.

tbg commented:
TFTRs!

bors r+
